browser/src/Renderer/WebGL/WebGLRenderer.ts

```diff
@@ -0,0 +1,535 @@
+                    const glyph = this._atlas.getGlyph(char, variantIndex)
+                    const colorToUse = cell.foregroundColor || defaultForegroundColor || "white"
+                    const normalizedTextColor = this._colorNormalizer.normalizeColor(colorToUse)
+
+                    this.updateGlyphInstance(
+                        glyphCount,
+                        Math.round(x - glyph.variantOffset), // TODO maybe remove the variantOffset here?
+                        // x,
+                        y,
+                        glyph,
+                        normalizedTextColor,
+                    )
+
+                    glyphCount++
+                    // x += glyph.subpixelWidth
+                }
+                x += pixelRatioAdaptedFontWidth
+            }
+
+            // y += subpixelFontHeight + subpixelLinePadding
+            y += pixelRatioAdaptedFontHeight
+        }
+        this._atlas.uploadTexture()
+
+        return glyphCount
+    }
+
+    private updateGlyphInstance(
+        index: number,
+        x: number,
+        y: number,
+        glyph: WebGLGlyph,
+        color: Float32Array,
+    ) {
+        const startOffset = GLYPH_INSTANCE_FIELD_COUNT * index
+        // targetOrigin
+        this.glyphInstances[0 + startOffset] = x
+        this.glyphInstances[1 + startOffset] = y
+        // targetSize
+        this.glyphInstances[2 + startOffset] = glyph.width
+        this.glyphInstances[3 + startOffset] = glyph.height
+        // textColorRGBA
+        this.glyphInstances[4 + startOffset] = color[0]
+        this.glyphInstances[5 + startOffset] = color[1]
+        this.glyphInstances[6 + startOffset] = color[2]
+        this.glyphInstances[7 + startOffset] = color[3]
+        // atlasOrigin
+        this.glyphInstances[8 + startOffset] = glyph.textureU
+        this.glyphInstances[9 + startOffset] = glyph.textureV
+        // atlasSize
+        this.glyphInstances[10 + startOffset] = glyph.textureWidth
+        this.glyphInstances[11 + startOffset] = glyph.textureHeight
+    }
+
+    private drawGlyphInstances(glyphCount: number, viewportScaleX: number, viewportScaleY: number) {
+        this._gl.bindVertexArray(this.textBlendVAO)
+        this._gl.enable(this._gl.BLEND)
+
+        this._gl.useProgram(this.textBlendPass1Program)
+        this._gl.uniform2f(this.textBlendPass1ViewportScaleLocation, viewportScaleX, viewportScaleY)
+        this._gl.bindBuffer(this._gl.ARRAY_BUFFER, this.glyphInstancesBuffer)
+        this._gl.bufferData(this._gl.ARRAY_BUFFER, this.glyphInstances, this._gl.STREAM_DRAW)
+        this._gl.blendFuncSeparate(
+            this._gl.ZERO,
+            this._gl.ONE_MINUS_SRC_COLOR,
+            this._gl.ZERO,
+            this._gl.ONE,
+        )
+        this._gl.drawElementsInstanced(this._gl.TRIANGLES, 6, this._gl.UNSIGNED_BYTE, 0, glyphCount)
+
+        this._gl.useProgram(this.textBlendPass2Program)
```

Comment:
Was curious - are the two passes needed for subpixel anti-aliasing? Or for handling transparency / background?

Reply:
The two-pass approach is required to do correct blending of subpixel-antialiased text onto our destination canvas. Basically we are abusing the RGB output channels of the first pass to do some intermediate calculations and blending with the destination for us which we then use to calculate the final result in the second pass.

 I will try to find the article again that explained the equations in a nice way and post it here!